### DISCUSSION TITLE
A potential code problem with the cascader-panel component

### DISCUSSION THREAD

**Discussion by xxf1996** [element-plus/element-plus#4443] [Q&A] [answered]
## Background

I use the cascader-panel component in my project, and when this component uses more than 1000 cascading options, the rendering becomes very slow causing the page to lag, roughly blocking the main thread for 1000ms or more.

## Solution

So I looked at the source code of the component and found [a suspicious code](https://github.com/element-plus/element-plus/blob/95ba055c63da48e28e085e6dd4a0beecfe18f825/packages/components/cascader-panel/src/index.vue#L84), so I made a simple local modification (changing the Ref variable to a normal variable) and found that the rendering performance was greatly improved, with rendering of more than 1000 cascade options taking only 100-200ms (still a bit slow, but acceptable).

![image](https://user-images.githubusercontent.com/17797002/142368936-fb8f341b-90f3-4507-9506-81616d3eecda.png)

## Discussion

I was going to raise an issue, but found that it could not be reproduced on [codepen](https://codepen.io/xxf1996/pen/RwZLJBg); so here I just want to discuss **whether the Ref variable in the code there is necessary**, because my local change of this Ref variable to a normal variable did not affect the normal use of any function (**no need for responsive processing** was found for this variable), and solved my problem.

**Comment by emojiiii** [comment]
I think you're right
@sxzz  cc

**Reply by xxf1996** [reply]
So are there any current plans to improve this area, after all maintaining a copy of the component source code in the project is not really a sensible approach.

**Comment by btea** [comment, marked as answer]
If change to a normal variable, dynamic load nodes seems have some problem

**Reply by xxf1996** [reply]
Does "dynamic load nodes" refer to the request back-end loading option or the lazy loading method? I really haven't tested the lazy load case, but the asynchronous request option is fine.

**Reply by btea** [reply]
Yes. The lazy load test will fail. You can have a try.

**Reply by xxf1996** [reply]
> Yes. The lazy load test will fail. You can have a try.

Indeed, I just tried to run unit tests and two test cases failed with this change; it seems that the change cannot simply be made in this way, although it can be asserted that the variable store itself has no responsive dependencies, but the nodes stored inside it may be dependent on the code somewhere.

**Reply by btea** [reply]
https://github.com/element-plus/element-plus/blob/b66c33dfacea3880a0a2c36c4b82de95d8158538/packages/components/cascader-panel/src/index.vue#L106-L108

It should be fine to adjust the code in this position like below.
```js
lazyLoad(undefined, (dataList) => {
    store = new Store(dataList, cfg)
    menus.value = [store.getNodes()]
    initialLoaded = true
    ...
})
```

**Reply by xxf1996** [reply]
What a coincidence, yesterday I also tried to reverse the code logic from the test case; is to locate this lazy load initialization place, after all, from the test case to see the initialization of the menus length is 0, indicating that the menus here is not updated, so try to manually update the menus in this lazyLoad callback, I adjusted the code yesterday and your very similar, as follows:

```js
lazyLoad(undefined, () => {
    menus.value = [store.getNodes()]
    initialLoaded = true
    ...
})
```

However, this modification did not pass the test, so I will give up this idea for now.😂

**Reply by emojiiii** [reply]
#4544
